### DISCUSSION TITLE
__activemask() question

### DISCUSSION THREAD

**Discussion by zjin-lcf** [intel/llvm#4755] [General]
Some cuda kernels call the __activemask() function. Is a SYCL function needed for that ?   Thanks.

https://github.com/mgawan/GPU-BSW/blob/master/kernel.cpp

__activemask():
Returns a 32-bit integer mask of all currently active threads in the calling warp. The Nth bit is set if the Nth lane in the warp is active when __activemask() is called. Inactive threads are represented by 0 bits in the returned mask.

**Comment by Michoumichmich** [comment]
Hello @zjin-lcf, I have looked at that in the past (and writing about the issue in my internship report ahah). There is https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc which could be intended to bridge the gap. I saw the implementation being committed some weeks ago.
However the notion of active threads might not really be defined in SYCL, see: https://github.com/KhronosGroup/SYCL-Docs/issues/193

**Reply by Michoumichmich** [reply]
The difficulty might also come from the fact that you can't return a 32-bit word. What is some GPU had warps with 64 threads ? or 128 ?

**Reply by zjin-lcf** [reply]
The authors ported their application with SYCL: https://github.com/mgawan/ADEPT/blob/sycl/adept/kernel.cpp

**Reply by Michoumichmich** [reply]
Sure, but they are not using a SYCL version of `__activemask()`, they worked around the issue. They are using sub-group algorithms which have to be called by the whole subgroup (so we can expect the whole warp to be active ?). The purpose of active mask was probably to allow using cuda primitives in case of control flow divergence... which isn't defined in SYC..
Maybe I missed your point, I don't understand the issue

**Reply by zjin-lcf** [reply]
No, the application with SYCL was redesigned. Developers will look forward to the SYCL version of __activemask().  Thanks.
